builtin/logical/pki/crl_util.go

```diff
@@ -1,24 +1,92 @@
 package pki
 
 import (
+	"bytes"
 	"context"
 	"crypto/rand"
 	"crypto/x509"
 	"crypto/x509/pkix"
 	"errors"
 	"fmt"
+	"math/big"
 	"strings"
+	"sync"
+	"sync/atomic"
 	"time"
 
+	"github.com/hashicorp/vault/sdk/helper/consts"
+
 	"github.com/hashicorp/vault/sdk/helper/certutil"
 	"github.com/hashicorp/vault/sdk/helper/errutil"
 	"github.com/hashicorp/vault/sdk/logical"
 )
 
+const revokedPath = "revoked/"
+
 type revocationInfo struct {
 	CertificateBytes  []byte    `json:"certificate_bytes"`
 	RevocationTime    int64     `json:"revocation_time"`
 	RevocationTimeUTC time.Time `json:"revocation_time_utc"`
+	CertificateIssuer issuerID  `json:"issuer_id"`
+}
+
+// crlBuilder is gatekeeper for controlling various read/write operations to the storage of the CRL.
+// The extra complexity arises from secondary performance clusters seeing various writes to its storage
+// without the actual API calls. During the storage invalidation process, we do not have the required state
+// to actually rebuild the CRLs, so we need to schedule it in a deferred fashion. This allows either
+// read or write calls to perform the operation if required, or have the flag reset upon a write operation
+type crlBuilder struct {
+	m            sync.Mutex
+	forceRebuild uint32
+}
+
+const (
+	_ignoreForceFlag  = true
+	_enforceForceFlag = false
+)
+
+// rebuildIfForced is to be called by readers or periodic functions that might need to trigger
+// a refresh of the CRL before the read occurs.
+func (cb *crlBuilder) rebuildIfForced(ctx context.Context, b *backend, request *logical.Request) error {
+	if atomic.LoadUint32(&cb.forceRebuild) == 1 {
+		return cb._doRebuild(ctx, b, request, true, _enforceForceFlag)
+	}
+
+	return nil
+}
+
+// rebuild is to be called by various write apis that know the CRL is to be updated and can be now.
+func (cb *crlBuilder) rebuild(ctx context.Context, b *backend, request *logical.Request, forceNew bool) error {
+	return cb._doRebuild(ctx, b, request, forceNew, _ignoreForceFlag)
+}
+
+// requestRebuildIfActiveNode will schedule a rebuild of the CRL from the next read or write api call assuming we are the active node of a cluster
+func (cb *crlBuilder) requestRebuildIfActiveNode(b *backend) {
```

Comment:
Can we restructure to avoid this problem? I think if we use `forceRebuild` as a true atomic, we could avoid holding the lock when _setting_ it and only grab the lock when reading it (inside `_doRebuild`). I think we can also change line 82 from `defer atomic.Store...` into a straight `atomic.Store` and the lock will prevent us from rebuilding the CRL at the same time -- the in-flight CRL build will be "wrong" if we've been requested to rebuild it for some reason anyways.

Reply:
Reworked in #15325